src/librustc_mir/transform/cs_propagate.rs

```diff
@@ -0,0 +1,425 @@
+//! | Join           | intersect | intersect |
+//!
+//! For all of them we will be using a lattice of `HashMap<Lvalue, Either<Lvalue, Constant, Top>>`.
+
+use rustc_data_structures::fnv::FnvHashMap;
+use rustc::middle::const_val::ConstVal;
+use rustc::mir::repr::*;
+use rustc::mir::tcx::binop_ty;
+use rustc::mir::transform::{Pass, MirPass, MirSource};
+use rustc::mir::visit::{MutVisitor};
+use rustc::ty::TyCtxt;
+use rustc_const_eval::{eval_const_binop, eval_const_unop, cast_const};
+use std::collections::hash_map::Entry;
+
+use super::dataflow::*;
+
+pub struct CsPropagate;
+
+impl Pass for CsPropagate {}
+
+impl<'tcx> MirPass<'tcx> for CsPropagate {
+    fn run_pass<'a>(&mut self, tcx: TyCtxt<'a, 'tcx, 'tcx>, _: MirSource, mir: &mut Mir<'tcx>) {
+        *mir = Dataflow::forward(mir, CsTransfer { tcx: tcx },
+                                 ConstRewrite { tcx: tcx }.and_then(SimplifyRewrite));
+    }
+}
+
+#[derive(PartialEq, Debug, Clone)]
+enum Either<'tcx> {
+    Lvalue(Lvalue<'tcx>),
+    Const(Constant<'tcx>),
+    Top
+}
+
+impl<'tcx> Lattice for Either<'tcx> {
+    fn bottom() -> Self { unimplemented!() }
+    fn join(&mut self, other: Self) -> bool {
+        if !other.eq(self) {
+            if Either::Top.eq(self) {
+                false
+            } else {
+                *self = Either::Top;
+                true
+            }
+        } else {
+            false
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct CsLattice<'tcx> {
+    values: FnvHashMap<Lvalue<'tcx>, Either<'tcx>>
+}
+
+impl<'tcx> CsLattice<'tcx> {
+    fn insert(&mut self, key: &Lvalue<'tcx>, val: Either<'tcx>) {
+        // FIXME: HashMap has no way to insert stuff without cloning the key even if it exists
+        // already.
+        match *key {
+            // Do not bother with statics – global state.
+            Lvalue::Static(_) => {}
+            // I feel like this could be handled, but needs special care. For example in code like
+            // this:
+            //
+            // ```
+            // var.field = false;
+            // something(&mut var);
+            // assert!(var.field);
+            // ```
+            //
+            // taking a reference to var should invalidate knowledge about all the
+            // projections of var and not just var itself. Currently we handle this by not
+            // keeping any knowledge about projections at all, but I think eventually we
+            // want to do so.
+            Lvalue::Projection(_) => {},
+            _ => match self.values.entry(key.clone()) {
+                Entry::Vacant(e) => {
+                    e.insert(val);
+                }
+                Entry::Occupied(mut e) => {
+                    e.get_mut().join(val);
+                }
+            }
+        }
+    }
+    fn remove(&mut self, key: &Lvalue<'tcx>) -> Option<Either<'tcx>> {
+        self.values.remove(key)
+    }
+    fn top(&mut self, key: &Lvalue<'tcx>) {
+        self.insert(key, Either::Top);
```

Comment:
Why not just remove from the map? With an intersection based join function, not present means `Top`.